torch/csrc/autograd/python_hook.h

```diff
@@ -34,5 +34,21 @@ struct PyFunctionPostHook : public FunctionPostHook {
   PyObject* dict;
 };
 
+// PyFunctionTensorPostAccGradHook is a DICTIONARY of PostAccumulateGradHooks,
+// NOT just one hook! Thus, adding one hook is actually just replacing the
+// existing dictionary with a (bigger) dictionary of hooks.
```

Comment:
Does the replace case actually happen? Well a contrived way of getting the hook replaced is to manually set the _backward_hooks attribute on a tensor but in the usual mode of operation when you add hooks from python you should just be appending to the existing dict?

Reply:
the replace seems to happen in variable.cpp code...

Reply:
But that's just the setter for _backward_hooks right? When you add a new hook through the `register_post_accumulate_grad_hook` logic, you only call that setter if _backward_hooks is None.

Reply:
Oh yea, but like basically the CPP side always replaces whereas the python dictionary gets added to